File: AdventOfCode/2025/Day4/Part2.cpp
#include <bits/stdc++.h>
#include <fstream>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    ifstream file("input.txt");
    if (!file.is_open()) return cerr << "ERROR", 1;

    vector<string> grid;
    string line;
    while (getline(file, line)) {
        if (line.empty()) continue;
        grid.push_back(line);
    }

    if (grid.empty()) {
        cout << 0 << '\n';
        return 0;
    }

    int rows = grid.size();
    int cols = grid[0].size();
    int totalremoved = 0;

    vector<int> dx = {-1, -1, -1, 0, 0, 1, 1, 1};
    vector<int> dy = {-1, 0, 1, -1, 1, -1, 0, 1};

    while (true) {
        vector<pair<int, int>> toremove;

        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                if (grid[i][j] != '@') continue;

                int count = 0;
                for (int d = 0; d < 8; ++d) {
                    int ni = i + dx[d];
                    int nj = j + dy[d];
                    if (ni >= 0 && ni < rows && nj >= 0 && nj < cols && grid[ni][nj] == '@') {
                        count++;
                    }
                }

                if (count < 4) {
                    toremove.push_back({i, j});
                }
            }
        }

        if (toremove.empty()) break;

        for (auto [i, j] : toremove) {
            grid[i][j] = '.';
        }

        totalremoved += toremove.size();
    }

    cout << totalremoved << '\n';
    return 0;
}

File: AdventOfCode/Day4/Python version/Part2.py
try:
    with open("input.txt", "r") as f:
        lines = []
        for line in f:
            stripped = line.rstrip('\n')
            if stripped != "":
                lines.append(stripped)
except FileNotFoundError:
    print("ERROR")
    exit()

if not lines:
    print(0)
    exit()



grid = [list(line) for line in lines]
rows = len(grid)
column = len(grid[0])
accessible = 0
total_removed = 0
directions = [(-1, -1), (-1, 0), (-1, 1),(0, -1), (0, 1), (1, -1),  (1, 0),  (1, 1)]

while True:
    toremove = []

    for i in range(rows):
        for j in range(column):
            if grid[i][j] == '@':
                count = 0
                for dx, dy in directions:
                    ni = i + dx
                    nj = j + dy
                    if 0 <= ni < rows and 0 <= nj < column:
                        if grid[ni][nj] == '@':
                            count += 1
                if count < 4:
                    toremove.append((i, j))

    if not toremove:
        break

    for (i, j) in toremove:
        grid[i][j] = '.'

    total_removed += len(toremove)

print(total_removed)